C/STrace/EtwLogger.h

```diff
@@ -0,0 +1,314 @@
+
+	return STATUS_SUCCESS;
+}
+
+// final recursion case
+template<typename FieldValue>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, FieldValue& fieldValue)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = ExAllocatePoolWithTag(NonPagedPoolNx, sizeof(FieldValue), 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	memcpy(newSpace, &fieldValue, sizeof(FieldValue));
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, sizeof(FieldValue));
+
+	return STATUS_SUCCESS;
+}
+
+// string-specific overload
+template<typename... Rest>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, const char* fieldValue, Rest... rest)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = (char*)ExAllocatePoolWithTag(NonPagedPoolNx, strlen(fieldValue) + 1, 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	strcpy(newSpace, fieldValue);
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, (ULONG)(strlen(fieldValue) + 1));
+
+	// Create the next descriptor with the next field value.
+	return CreateTracePropertyRecursive(++propertyDataDescriptors, rest...);
+}
+
+template<typename FieldValue, typename... Rest>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, FieldValue& fieldValue, Rest... rest)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = ExAllocatePoolWithTag(NonPagedPoolNx, sizeof(FieldValue), 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	memcpy(newSpace, &fieldValue, sizeof(FieldValue));
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, sizeof(FieldValue));
+
+	// Create the next descriptor with the next field value.
+	return CreateTracePropertyRecursive(++propertyDataDescriptors, rest...);
+}
+
+EVENT_DESCRIPTOR CreateEventDescriptor(uint64_t keyword, uint8_t level)
+{
+	EVENT_DESCRIPTOR desc;
+	RtlSecureZeroMemory(&desc, sizeof(EVENT_DESCRIPTOR));
+	desc.Channel = 11;  // All "manifest-free" events should go to channel 11 by default
```

Comment:
The [original interface](https://github.com/microsoft/DTrace-on-Windows/blob/windows/samples/windows/etw/AddNewETWEvent.d#L31) doesn't let you set it, and [this](https://learn.microsoft.com/en-us/windows/win32/api/evntprov/ns-evntprov-event_descriptor) documentation seems to say the channel is fixed to specific values (11 is for TraceLogging, which I guess we're emulating since we don't have a manifest?).

It's not hard to add it to our interface though if you want.

Reply:
lets add it just for fun, I have no idea if it's useful but why not. We can just make it default to 11